### ISSUE TITLE
[request] PHP (summary issue)

### ISSUE THREAD

**Issue by friendlyanon** [conan-io/conan-center-index#8618]
This is a summary issue tracking progress for dependencies of the [php-src](https://github.com/php/php-src) runtime of PHP programming language. These are provided on [windows.php.net](https://windows.php.net/downloads/php-sdk/deps/vs16/x64/) for Windows, but the goal would be to make them build easily on all platforms in a reproducible way.

### Not on CCI yet or needs some work

- apache (https://httpd.apache.org/)
- c-client (https://github.com/uw-imap/imap)
- fbclient (https://www.firebirdsql.org/pdfmanual/html/ufb-cs-clientlib.html)
- glib (https://www.gtk.org/)
  - Can't build as shared library on Windows.
- ICU (https://icu.unicode.org/)
  - Some platforms aren't supported. Is this an issue?
- libargon2 (https://github.com/p-h-c/phc-winner-argon2)
  - There is already libsodium. Is this needed?
- libavif (https://github.com/AOMediaCodec/libavif)
- libenchant2 (https://github.com/AbiWord/enchant)
- libffi (https://sourceware.org/libffi/)
  - Package info needs better CMake support.
- libintl (https://www.gnu.org/software/gettext/gettext.html)
  - Seems to be part of gettext. Is that right?
- liblmdb (https://www.symas.com/lmdb)
  - Package info needs better CMake support.
- libqdbm (http://fallabs.com/qdbm/)
- libsasl (https://www.cyrusimap.org/sasl/)
  - Doesn't compile on Windows.
- libsodium (https://github.com/jedisct1/libsodium)
  - Can't build with VS <2015 and shared library with static runtime doesn't work. Is this a problem?
- libxpm (https://cgit.freedesktop.org/xorg/lib/libXpm)
- libxslt (https://xmlsoft.org)
  - Package info needs better CMake support.
- net-snmp (http://www.net-snmp.org/)
- nghttp2 (https://nghttp2.org)
  - Package info needs better CMake support.
- openldap (https://www.openldap.org/)
  - Basically finished locally, just needs Cyrus SASL as a dependency.
- wineditline (http://mingweditline.sourceforge.net)
  - #8553

I will keep editing this list as I work through each item. If I got something wrong, then please let me know.

**Comment by friendlyanon** [comment]
Mailing list post: https://externals.io/message/116655#116779

**Comment by pierrejoye** [comment]
glib

we do build it shared already for enchant. What is the issue you are experiencing?

libintl

yes, part of gettext

**Comment by friendlyanon** [comment]
> What is the issue you are experiencing?

https://github.com/conan-io/conan-center-index/blob/58a0fe0fcc0f9daaeaad35b22f75bc2c70e0d14c/recipes/glib/all/conanfile.py#L52-L56

**Comment by pierrejoye** [comment]
ah, that's no issue, we use the shared build only.

**Comment by friendlyanon** [comment]
Okay, so it seems there is a bit of a cyclic dependency going on with Cyrus SASL and OpenLDAP.

The OpenLDAP project has a component called lmdb, which is necessary to build the sasldb plugin for Cyrus SASL and then OpenLDAP depends on Cyrus SASL for full functionality. Upstream Cyrus SASL also seems to be using some weird conan setup and ships VS solutions for Windows in the `win32` folder, so that could be used on Windows hopefully.

There is also the ldapdb plugin, which I'm not sure is useful.  
It seems that lmdb must be broken out into its own recipe.

cc @uilianries

**Comment by friendlyanon** [comment]
A convenient solution to libffi having a custom patch necessary for use in PHP is to use [Barbarian](https://barbarian.bfgroup.xyz/).  
This would allow us to stay in the Conan ecosystem without having to host an Artifactory server for custom recipes.

**Comment by Croydon** [comment]
> A convenient solution to libffi having a custom patch necessary for use in PHP is to use [Barbarian](https://barbarian.bfgroup.xyz/). This would allow us to stay in the Conan ecosystem without having to host an Artifactory server for custom recipes.

Unfortunately, this is not a solution as CCI recipes can only depend on other CCI recipes

**Comment by friendlyanon** [comment]
The goal is not to have php-src in CCI, but to have all its dependencies on CCI, so it can be built easily on all platforms once it has a CMake build tooling. For php-src, that's perfectly fine, since no other dependency has libffi as a dependency.

For php-src to be available on CCI as well, all of its dependencies that are patched and can't be upstreamed for various reasons would first have to be in a desirable condition upstream. Let's just say I don't want to go down that rabbit hole myself and Barbarian is a good enough solution in the meantime.

**Comment by Croydon** [comment]
> The goal is not to have php-src in CCI, but to have all its dependencies on CCI, so it can be built easily on all platforms once it has a CMake build tooling. For php-src, that's perfectly fine, since no other dependency has libffi as a dependency.

Understood. I find it personally interesting to eventually have PHP itself in CCI too, but let's keep this out of this issue for now. Perhaps change the title of this issue to something like `Dependencies of PHP`?

Huge 👍 for working towers a CMake + Conan build system for PHP, btw. 👏

**Comment by friendlyanon** [comment]
I looked at libxpm and some things aren't quite clear to me. The [`Makefile.am`](https://gitlab.freedesktop.org/xorg/lib/libxpm/-/blob/master/src/Makefile.am) contained in the upstream repository contains more sources for the library than the [PHP build files for Windows](https://github.com/winlibs/libxpm/blob/master/windows/vs16/libxpm.vcxproj#L312-L332). Additionally, the windows.php.net distribution of the library contains these additional header files:

* keysym.h
* X.h
* Xfuncproto.h
* Xlib.h
* Xosdefs.h
* Xutil.h

I'm told these files are in the distribution only because the `ext/gd` extension for PHP depends on those headers. This seems like an improper dependency hierarchy to me.

**Comment by friendlyanon** [comment]
@cmb69 I looked at the `ext/gd` directory in php-src and it appears that none of the headers are used directly. Are you sure these aren't an artifact of decisions that aren't relevant anymore?

**Comment by cmb69** [comment]
[xpm.h includes X11/Xlib.h and X11/Xutil.h](https://github.com/winlibs/libxpm/blob/76184718821c4788ab0cfc4ce6fc153cddebd917/include/X11/xpm.h#L76-L77). If `FOR_MSW` was defined, simx.h would be included, but that is apparently not available either. The clean solution would likely be to rely on another package, but I presume that X11 can't be built under native Windows. On the other hand, I'm not sure that anybody uses XBM files under Windows.

**Comment by pierrejoye** [comment]
@friendlyanon gd.h and a few are used during build.

Other may be used only when builds are done against ext/gd or its libgd.

As of libgd itself, you can check https://github.com/libgd/libgd as it has cmake support already.

**Comment by friendlyanon** [comment]
The recipe in #8786 installs only xpm.h and simx.h, see [here](https://github.com/conan-io/conan-center-index/pull/8786/files#diff-d5a8533e30c1d03ece41b3c07ad32b3201b53473e7e362527680af809314042eR91-R95).

It also defines the `FOR_MSW` header for consumers [here](https://github.com/conan-io/conan-center-index/pull/8786/files#diff-3962d1c19920835936bcdb5f9d4f308f0cbe0b518520c5bd4b1b5097b68986efR80-R81). It's likely that PHP was missing this define during build and thus you still needed the headers, despite building on Windows.

**Comment by cmb69** [comment]
@pierrejoye, PHP on Windows still doesn't support building against external libgd, and I'm reluctant to introduce that until custom memory allocators are available.

@friendlyanon, ah, indeed, thank you! I'll roll out a new libxpm package with this change soon.

**Comment by pierrejoye** [comment]
@cmb69 correct. However the whole build can be a lot easier by taking the what is needed from libgd/libgd's cmake. Made a static build and uses it for ext/gd. It may requires a few tweaks but as it is built at the same time than php, it should not require much changes.

**Comment by friendlyanon** [comment]
So, I have been trying to crack the nut that is the Apache HTTP Server and boy, does it have a million knobs. (also unsatisfactory Windows support in build tooling aspect)

I looked at the php-src codebase and a grep for `apache` doesn't immediately tell me what kind of featureset PHP expects from the HTTP library. https://github.com/winlibs/apache has also been abandoned for 7 years now, so that is no use here either. I tried looking at what features Linux distributions enable, and I found at least the [Alpine build script](https://git.alpinelinux.org/aports/tree/main/apache2/APKBUILD#n142).

@cmb69 @pierrejoye

**Comment by cmb69** [comment]
@friendlyanon, we using the [Apache Lounge builds of Apache](https://www.apachelounge.com/download/) for years. I have no idea, how they do these builds exactly, but could ask, if necessary.
